### PR TITLE
PWN-9472 -  Dynamic percent diff

### PR DESCRIPTION
**Pull request by Davran331122**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-9472

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/user/interactor/UserTokensInteractor.kt [hunk 1]
```diff
@@ -55,6 +56,10 @@ class UserTokensInteractor(
         return userTokensLocalRepository.observeUserTokens()
     }
 
+    fun observeUserTokens(mintAddress: Base58String): Flow<Token.Active> {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/user/interactor/UserTokensInteractor.kt:59]: observeUserToken - not tokens

File: app/src/main/java/org/p2p/wallet/user/interactor/UserTokensInteractor.kt [hunk 2]
```diff
@@ -68,4 +73,11 @@ class UserTokensInteractor(
     suspend fun getUserBalance(): BigDecimal {
         return userTokensLocalRepository.getUserBalance()
     }
+
+    suspend fun setTokenHidden(mintAddress: String, visibility: String) =
+        userTokensLocalRepository.setTokenHidden(mintAddress, visibility)
+
+    suspend fun clear() {
+        userTokensLocalRepository.clear()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/user/interactor/UserTokensInteractor.kt:81]: we clear repositories, but not interactors, it's better to inject userTokensLocalRepository and clear it

File: app/src/main/java/org/p2p/wallet/user/interactor/UserTokensInteractor.kt [hunk 3]
```diff
@@ -68,4 +73,11 @@ class UserTokensInteractor(
     suspend fun getUserBalance(): BigDecimal {
         return userTokensLocalRepository.getUserBalance()
     }
+
+    suspend fun setTokenHidden(mintAddress: String, visibility: String) =
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/user/interactor/UserTokensInteractor.kt:77]: i think there should be a separate interactor like HiddenTokenInteractor for this logic

File: app/src/main/java/org/p2p/wallet/user/repository/UserTokensDatabaseRepository.kt [hunk 4]
```diff
@@ -53,13 +53,19 @@ class UserTokensDatabaseRepository(
...
+            }
     }
 
     override fun observeUserTokens(): Flow<List<Token.Active>> {
         return tokensDao.getTokensFlow()
             .map { tokenEntities ->
-                tokenEntities.map(tokenConverter::fromDatabase)
+                tokenEntities.map { tokenEntity ->
+                    val tokenData = userLocalRepository.findTokenData(tokenEntity.mintAddress)
+                    tokenConverter.fromDatabase(tokenEntity, tokenData?.extensions)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/user/repository/UserTokensDatabaseRepository.kt:67]: duplication 3 times

File: buildSrc/src/main/java/AppVersions.kt [hunk 5]
```diff
@@ -1,7 +1,7 @@
 object AppVersions {
     const val VERSION_MAJOR = 2
-    const val VERSION_MINOR = 9
```
Comment (gslevinkov) [inline buildSrc/src/main/java/AppVersions.kt:3]: revert

File: token-service/src/main/java/org/p2p/token/service/database/TokenPriceDao.kt [hunk 6]
```diff
@@ -20,10 +20,10 @@ internal interface TokenPriceDao {
     suspend fun getAllTokensPrices(): List<TokenServicePriceEntity>
 
     @Query("SELECT * FROM token_price_table")
-    fun getAllTokensPricesFlow(): Flow<TokenServicePriceEntity>
+    fun getAllTokensPricesFlow(): Flow<List<TokenServicePriceEntity>>
 
-    @Query("SELECT * FROM token_price_table WHERE token_address = :tokenAddress")
-    suspend fun getTokenPriceByAddress(tokenAddress: String): TokenServicePriceEntity?
+    @Query("SELECT * FROM token_price_table WHERE token_column_key = :tokenColumnKey")
```
Comment (gslevinkov) [inline token-service/src/main/java/org/p2p/token/service/database/TokenPriceDao.kt:25]: why the previous logic was not working?

File: token-service/src/main/java/org/p2p/token/service/repository/price/TokenPriceDatabaseRepository.kt [hunk 7]
```diff
@@ -18,9 +18,11 @@ internal class TokenPriceDatabaseRepository(
     }
 
     override suspend fun findTokenPriceByAddress(
-        address: String
+        address: String,
+        networkChain: TokenServiceNetwork
     ): TokenServicePrice? {
-        val found = tokenPriceDao.getTokenPriceByAddress(address)
+        val tokenColumnKey = "${address}_${networkChain.networkName}"
```
Comment (gslevinkov) [inline token-service/src/main/java/org/p2p/token/service/repository/price/TokenPriceDatabaseRepository.kt:24]: the generation logic for this column should be extracted

File: app/src/main/java/org/p2p/wallet/user/interactor/UserTokensInteractor.kt (hunk 3 above)
Comment (Davran331122) [reply app/src/main/java/org/p2p/wallet/user/interactor/UserTokensInteractor.kt:77]: maybe ) let's do this in next )

File: token-service/src/main/java/org/p2p/token/service/database/TokenPriceDao.kt (hunk 6 above)
Comment (Davran331122) [reply token-service/src/main/java/org/p2p/token/service/database/TokenPriceDao.kt:25]: we have collision with native eth and native sol tokens rates

File: buildSrc/src/main/java/AppVersions.kt [hunk 8]
```diff
@@ -1,6 +1,6 @@
 object AppVersions {
     const val VERSION_MAJOR = 2
-    const val VERSION_MINOR = 8
+    const val VERSION_MINOR = 9
     const val VERSION_HOTFIX = 0
     const val VERSION_PATCH = 2
```
Comment (gslevinkov) [inline buildSrc/src/main/java/AppVersions.kt:5]: please check the AppVersion at release/2.9.0, now it's not correct

File: token-service/src/main/java/org/p2p/token/service/repository/price/TokenPriceDatabaseRepository.kt [hunk 9]
```diff
@@ -36,3 +36,7 @@ internal class TokenPriceDatabaseRepository(
             .map(converter::fromEntity)
     }
 }
+
+internal fun String.toPriceColumnKey(networkChainName: String): String {
```
Comment (gslevinkov) [inline token-service/src/main/java/org/p2p/token/service/repository/price/TokenPriceDatabaseRepository.kt:40]: and put it in the Entity too
that's why we are extracting this logic to a method - to remove errors if it's changed

**Review by gslevinkov — APPROVED** [review]

**Pull request merged**